src/app/modules/submission/encrypt-submission/encrypt-submission.controller.ts

```diff
@@ -94,286 +85,258 @@ const Payment = getPaymentModel(mongoose)
-  SubmissionResponseDto | SubmissionErrorDto,
-  EncryptSubmissionDto,
-  { captchaResponse?: unknown }
->
+const submitEncryptModeForm: SharedSubmissionMiddlewareHandlerType = async (
+  req,
+  res,
+) => {
+  const { formId } = req.params
 
-const submitEncryptModeForm: SubmitEncryptModeFormControllerHandlerType =
-  async (req, res) => {
-    const { formId } = req.params
+  if ('isPreview' in req.body) {
+    logger.info({
+      message:
+        'isPreview is still being sent when submitting encrypt mode form',
+      meta: {
+        action: 'submitEncryptModeForm',
+        type: 'deprecatedCheck',
+      },
+    })
+  }
 
-    if ('isPreview' in req.body) {
-      logger.info({
-        message:
-          'isPreview is still being sent when submitting encrypt mode form',
-        meta: {
-          action: 'submitEncryptModeForm',
-          type: 'deprecatedCheck',
-        },
-      })
-    }
+  const logMeta = {
+    action: 'submitEncryptModeForm',
+    ...createReqMeta(req),
+    formId,
+  }
 
-    const logMeta = {
-      action: 'submitEncryptModeForm',
-      ...createReqMeta(req),
-      formId,
-    }
+  const formDef = req.body.formsg.formDef
+  if (!formDef) return res.send(new FormDefinitionNotRetrievedError())
 
-    // Retrieve form
-    const formResult = await FormService.retrieveFullFormById(formId)
-    if (formResult.isErr()) {
-      logger.warn({
-        message: 'Failed to retrieve form from database',
-        meta: logMeta,
-        error: formResult.error,
-      })
-      const { errorMessage, statusCode } = mapRouteError(formResult.error)
-      return res.status(statusCode).json({ message: errorMessage })
-    }
+  setFormTags(formDef)
 
-    setFormTags(formResult.value)
+  const form = req.body.formsg.encryptedFormDef
+  if (!form) return res.send(new EncryptedFormDefinitionNotRetrievedError())
 
-    const checkFormIsEncryptModeResult = checkFormIsEncryptMode(
-      formResult.value,
-    )
-    if (checkFormIsEncryptModeResult.isErr()) {
-      logger.error({
-        message:
-          'Trying to submit non-encrypt mode submission on encrypt-form submission endpoint',
-        meta: logMeta,
-      })
-      const { statusCode, errorMessage } = mapRouteError(
-        checkFormIsEncryptModeResult.error,
-      )
-      return res.status(statusCode).json({
-        message: errorMessage,
-      })
-    }
-    const form = checkFormIsEncryptModeResult.value
+  const ensurePipeline = new Pipeline(
+    ensurePublicForm,
+    ensureValidCaptcha,
+    ensureFormWithinSubmissionLimits,
+  )
 
-    const ensurePipeline = new Pipeline(
-      ensurePublicForm,
-      ensureValidCaptcha,
-      ensureFormWithinSubmissionLimits,
-    )
+  const hasEnsuredAll = await ensurePipeline.execute({
+    form,
+    logMeta,
+    req,
+    res,
+  })
 
-    const hasEnsuredAll = await ensurePipeline.execute({
-      form,
-      logMeta,
-      req,
-      res,
-    })
+  if (!hasEnsuredAll) return
```

Comment:
clarification as first time reviewing the middleware - what happens if one for the middleware fails, will the correct status and message be sent in the response?

Reply:
The ensures are responsible for sending out the response and failing the pipeline. So they **must** do both as convention for this particular sets of ensures. I have not seen other use cases of ensures yet, thus didn't require that all ensures to follow such convention. I suppose that once we have more datapoints, we can make a better call whether to have this as a convention.

Reply:
Actually yea related question, will the line `if (!hasEnsuredAll) return` ever enter the loop? Because I'd imagine that if any of the ensures fail, it will send the response directly. If anything fails and a response is not sent, the statement implies that there will not be a response sent, so the client will just be waiting forever.

Reply:
Maybe we could add a catchall res here? note that res.status will send the latest, so we should check it has not been set first

Reply:
Refactored to use a catch-all res instead in [64ca3e8](https://github.com/opengovsg/FormSG/pull/6595/commits/64ca3e8db22e4940d16c752e0a9a6c5d82430fe3)!